### PR TITLE
Add ConsistentDropout 

### PR DESCRIPTION
**Pull request by Dref360**
## Summary:

This (second) PR adds a new component for research user: ConsistentDropout.
While ConsistentDropout is a bit slower, it guarantees the same weights drawn from the posterior distribution for each batch. This is useful when doing research to give every sample the same distribution.

### Features:

It has the same API as baal.bayesian.Dropout so I won't cover it.

The behavior can be seen with this example:

```python
your_model = baal.bayesian.consistent_dropout.patch_module(YourModel())
model = ModelWrapper(your_model)

a_batch = torch.randn(1, 32)  # assuming an input size of 32

out = model.predict_on_batch(a_batch, iterations=10)
assert out.shape[-1] == 10  # We have 10 iterations.

# We compute the uncertainty per batch and it works!
assert (torch.var(out, -1) > 0).all()

# But because we draw the same weights, the preds are the same if we predict again
out2 = model.predict_on_batch(a_batch, iterations=10)
assert torch.eq(out, out2).all()
```

## Checklist:

* [x] Your code is documented (To validate this, add your module to `tests/documentation_test.py`).
* [x] Your code is tested with unit tests.
* [x] You moved your Issue to the PR state.

### CODE REVIEW

File: src/baal/bayesian/consistent_dropout.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    ConsistentDropout is useful when doing research.
+    It guarantees that while the mask are the same between batches,
+    they are different inside the batch.
+
+    This is slower than using regular Dropout, but it is useful
+    when you want to use the same set of weights for each unlabelled sample.
+
+    Args:
+        p (float): probability of an element to be zeroed. Default: 0.5
+    """
```
Comment (Dref360) [inline src/baal/bayesian/consistent_dropout.py:19]:
```suggestion
	
	Notes:
		For optimal results, you should use a batch size of one
		during inference time.
		Furthermore, to guarantee that each sample uses the same
		set of weights,
		you must use `replicate_in_memory=True` in ModelWrapper
		, which is the default.  
    """
```

File: src/baal/bayesian/consistent_dropout.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+import torch
+from torch.nn import functional as F
+from torch.nn.modules.dropout import _DropoutNd
+
+
+class ConsistentDropout(_DropoutNd):
+    """
+    ConsistentDropout is useful when doing research.
+    It guarantees that while the mask are the same between batches,
```
Comment (orangetoaster) [inline src/baal/bayesian/consistent_dropout.py:11]: during inference.

File: src/baal/bayesian/consistent_dropout.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    they are different inside the batch.
```
Comment (Dref360) [inline src/baal/bayesian/consistent_dropout.py:12]:
```suggestion
    It guarantees that while the masks are the same between batches
    during inference. The masks are different inside the batch.
```

File: src/baal/bayesian/consistent_dropout.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+    This is slower than using regular Dropout, but it is useful
+    when you want to use the same set of weights for each unlabelled sample.
```
Comment (Dref360) [inline src/baal/bayesian/consistent_dropout.py:15]:
```suggestion
    when you want to use the same set of weights for each sample used in inference.
    
    From BatchBALD (Kirsch et al, 2019), this is necessary to use BatchBALD and remove noise 
    from the prediction.
```

File: src/baal/bayesian/consistent_dropout.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    def forward(self, x):
+        if self.training:
+            return F.dropout(x, self.p, training=True, inplace=False)
+        else:
+            if self._mask is None or self._mask.shape != x.shape:
+                self._mask = self._make_mask(x)
+            return torch.mul(x, self._mask)
+
+    def _make_mask(self, x):
+        return F.dropout(torch.ones_like(x, device=x.device), self.p, training=True)
```
Comment (BlackHC) [inline src/baal/bayesian/consistent_dropout.py:45]: Could you set the batch dimension to 1 here, so it will broadcast?  (Or you can also drop it which should have the same effect IIRC.)

We want the masks to be consistent within the batch, too.

File: src/baal/bayesian/consistent_dropout.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    def forward(self, x):
+        if self.training:
+            return F.dropout2d(x, self.p, training=True, inplace=False)
+        else:
+            if self._mask is None or self._mask.shape != x.shape:
+                self._mask = self._make_mask(x)
+            return torch.mul(x, self._mask)
+
+    def _make_mask(self, x):
+        return F.dropout2d(torch.ones_like(x, device=x.device), self.p, training=True)
```
Comment (BlackHC) [inline src/baal/bayesian/consistent_dropout.py:81]: Same here.

File: src/baal/bayesian/consistent_dropout.py (hunk 5 above)
Comment (Dref360) [reply src/baal/bayesian/consistent_dropout.py:45]: (Just our discussion on Slack to have everything at the same place)
In the note, we said that we expected a batch size of 1 and to use replicate_in_mem=True.

I am trying to come up with something to remove this constraint or better show it. I'm open to suggestion :)
